diff --git a/Lib/JpGraph/4.2.4/gd_image.inc.php b/Lib/JpGraph/4.2.4/gd_image.inc.php
index 2074036..9bff950 100644
--- a/Lib/JpGraph/4.2.4/gd_image.inc.php
+++ b/Lib/JpGraph/4.2.4/gd_image.inc.php
@@ -108,7 +108,7 @@ class Image {
             imageantialias($this->img,$aFlg);
         }
         else {
-            JpGraphError::RaiseL(25128);//('The function imageantialias() is not available in your PHP installation. Use the GD version that comes with PHP and not the standalone version.')
+            // PATCH: skip this error // JpGraphError::RaiseL(25128);JpGraphError::RaiseL(25128);//('The function imageantialias() is not available in your PHP installation. Use the GD version that comes with PHP and not the standalone version.')
         }
     }
 
