Index: checksolution.php
===================================================================
--- checksolution.php	(revision 0)
+++ checksolution.php	(revision 0)
@@ -0,0 +1,243 @@
+<?php
+
+
+/***************************************************************************
+	*
+	*   This program is free software; you can redistribute it and/or modify
+	*   it under the terms of the GNU General Public License as published by
+	*   the Free Software Foundation; either version 2 of the License, or
+	*   (at your option) any later version.
+	*
+	***************************************************************************/
+
+/****************************************************************************
+
+   Unicode Reminder ąśćł
+
+*/
+
+  //prepare the templates and include all neccessary
+	require_once('./lib/common.inc.php');
+	require($stylepath.'/smilies.inc.php');
+
+	function getUserSolvedLog($cache_id)
+	{
+		global $usr;
+		
+		$log_id = 0;
+		$rs = sql("SELECT `id` FROM `cache_logs` WHERE `cache_id` = &1 AND `user_id` = &2 AND `type` = 9", $cache_id, $usr['userid']);
+
+		if ($r = sql_fetch_assoc($rs))
+		{
+			$log_id = $r['id'];
+		}
+
+		mysql_free_result($rs);
+
+		return $log_id;
+	}
+
+	//Preprocessing
+	if ($error == false)
+	{
+		//user logged in?
+		if ($usr == false)
+		{
+			$target = urlencode(tpl_get_current_page());
+			tpl_redirect('login.php?target='.$target);
+		}
+		else
+		{
+			$log_id = 0;
+
+			if (isset($_REQUEST['logid']))
+			{
+				$log_id = $_REQUEST['logid'];
+			}
+			else
+			{
+				$cacheid = $_GET['cacheid'];
+				$solution = $_GET['solution'];
+				$solution_rs = sql("SELECT `solution`, `status` FROM `caches` WHERE `cache_id` = &1", $cacheid);
+
+				if (mysql_num_rows($solution_rs) > 0)
+				{
+					$solution_r = sql_fetch_array($solution_rs);
+
+					if ($solution_r['status'] != 4 && $solution_r['status'] != 5 && $solution_r['status'] != 6)
+					{
+						$is_solved = $solution_r['solution'] == $solution;
+
+						if ($is_solved)
+						{
+							$log_type = 9;
+							$solved_column = 'solved';
+							tpl_set_var('solution_status', tr('correct_solution'));
+						}
+						else
+						{
+							$log_type = 10;
+							$solved_column = 'not_solved';
+							tpl_set_var('solution_status', tr('incorrect_solution'));
+						}
+
+						$log_id = getUserSolvedLog($cacheid);
+
+						if ($is_solved && $log_id > 0)
+						{
+							sql("UPDATE `cache_logs` SET `date`=NOW(), `last_modified`=NOW(), `text`=NULL, `text_html`=1, `text_htmledit`=1, `deleted`=0, `hidden`=1 WHERE id = &1", $log_id);
+						}
+						else
+						{
+							sql("UPDATE `caches` SET `&1` = `&1` + 1 WHERE `cache_id` = &2", $solved_column, $cacheid);
+
+							$log_uuid = create_uuid();
+							sql("INSERT INTO `cache_logs` (`id`, `cache_id`, `user_id`, `type`, `date`, `date_created`, `last_modified`, `uuid`, `node`, `text_html`, `text_htmledit`, `hidden`)
+									VALUES ('', '&1', '&2', '&3', NOW(), NOW(), NOW(), '&4', '&5', 1, 1, 1)",
+									$cacheid, $usr['userid'], $log_type, $log_uuid, $oc_nodeid);
+							
+							$log_rs = sql("SELECT id FROM cache_logs WHERE uuid = '&1'", $log_uuid);
+
+							if ($log_record = sql_fetch_assoc($log_rs))
+							{
+								$log_id = $log_record['id'];
+							}
+
+							mysql_free_result($log_rs);
+						}
+					}
+				}
+			}
+
+			if ($log_id > 0)
+			{
+				$log_rs = sql("SELECT `cache_logs`.`cache_id` AS `cache_id`, `cache_logs`.`node` AS `node`, `cache_logs`.`type` AS `logtype`, `caches`.`name` AS `cachename`, `caches`.`type` AS `cachetype`, `caches`.`user_id` AS `cache_user_id` FROM `cache_logs` INNER JOIN `caches` ON (`caches`.`cache_id`=`cache_logs`.`cache_id`) WHERE `id`='&1' AND `deleted` = &2", $log_id, 0);
+
+				if (mysql_num_rows($log_rs) > 0)
+				{
+					$log_record = sql_fetch_array($log_rs);
+					require($stylepath . '/checksolution.inc.php');
+					require_once($rootpath . 'lib/caches.inc.php');
+
+					if ($log_record['node'] != $oc_nodeid)
+					{
+						tpl_errorMsg('checksolution', $error_wrong_node);
+						exit;
+					}
+
+					$tplname = 'checksolution';
+
+					//load settings
+					$cache_name = $log_record['cachename'];
+					$cache_type = $log_record['cachetype'];
+					$cache_user_id = $log_record['cache_user_id'];
+
+					if (isset($_POST['descMode']))
+					{
+						$descMode = $_POST['descMode'] + 0;
+
+						if ($descMode < 1 || $descMode > 3)
+							$descMode = 3;
+					}
+					else
+					{
+						$descMode = 3;
+					}
+
+					if ($descMode != 1)
+					{
+						// Text from textarea
+						$log_text = isset($_POST['logtext']) ? ($_POST['logtext']) : '';
+
+						// check input
+						require_once($rootpath . 'lib/class.inputfilter.php');
+						$myFilter = new InputFilter($allowedtags, $allowedattr, 0, 0, 1);
+						$log_text = $myFilter->process($log_text);
+					}
+					else
+					{
+						// escape text
+						$log_text = isset($_POST['logtext']) ? htmlspecialchars($_POST['logtext'], ENT_COMPAT, 'UTF-8') : '';
+					}
+
+					if (isset($_POST['submitform']))
+					{
+						sql("UPDATE `cache_logs` SET `text`='&1', `text_html`='&2', `text_htmledit`='&3', `last_modified`=NOW(), `hidden`=0
+							WHERE `id`='&4'",
+							tidy_html_description((($descMode != 1) ? $log_text : nl2br($log_text))), (($descMode != 1) ? 1 : 0), (($descMode == 3) ? 1 : 0), $log_id);
+
+						require_once($rootpath . 'lib/eventhandler.inc.php');
+						event_change_log_type($log_record['cache_id'], $usr['userid']);
+					}
+
+					if (isset($_POST['submitform']) || isset($_POST['cancelform']))
+					{
+						tpl_redirect('viewcache.php?cacheid=' . urlencode($log_record['cache_id']));
+						exit;
+					}
+
+					//set template vars
+					tpl_set_var('cachename', htmlspecialchars($cache_name, ENT_COMPAT, 'UTF-8'));
+					tpl_set_var('cachename', htmlspecialchars($cache_name, ENT_COMPAT, 'UTF-8'));
+					tpl_set_var('cacheid', $log_record['cache_id']);
+					tpl_set_var('logid', $log_id);
+
+					$log_text = tidy_html_description($log_text);
+
+					if ($descMode != 1)
+						tpl_set_var('logtext', htmlspecialchars($log_text, ENT_NOQUOTES, 'UTF-8'), true);
+					else
+						tpl_set_var('logtext', $log_text);
+
+					// Text / normal HTML / HTML editor
+					tpl_set_var('use_tinymce', (($descMode == 3) ? 1 : 0));
+
+					if ($descMode == 1)
+						tpl_set_var('descMode', 1);
+					else if ($descMode == 2)
+						tpl_set_var('descMode', 2);
+					else
+					{
+						// TinyMCE
+						$headers = tpl_get_var('htmlheaders') . "\n";
+						$headers .= '<script language="javascript" type="text/javascript" src="lib/phpfuncs.js"></script>' . "\n";
+						$headers .= '<script language="javascript" type="text/javascript" src="lib/tinymce/tiny_mce.js"></script>' . "\n";
+						$headers .= '<script language="javascript" type="text/javascript" src="lib/tinymce/config/log.js.php?lang='.$lang.'&amp;dateformat='.$dateformat.'&amp;logid=0"></script>' . "\n";
+						tpl_set_var('htmlheaders', $headers);
+						tpl_set_var('descMode', 3);
+					}
+
+					// build smilies
+					$smilies = '';
+					if ($descMode != 3)
+					{
+						for($i=0; $i<count($smileyshow); $i++)
+						{
+							if($smileyshow[$i] == '1')
+							{
+								$tmp_smiley = $smiley_link;
+								$tmp_smiley = mb_ereg_replace('{smiley_image}', $smileyimage[$i], $tmp_smiley);
+								$smilies = $smilies.mb_ereg_replace('{smiley_text}', ' '.$smileytext[$i].' ', $tmp_smiley).'&nbsp;';
+							}
+						}
+					}
+					tpl_set_var('smilies', $smilies);
+				}
+				else
+				{
+					// no such log or log marked as deleted
+					header('HTTP/1.0 404 not found');
+					include('./error_pages/404.html');
+					exit();
+				}
+			}
+			else
+			{
+				header('Location: viewcache.php?cacheid='.$cacheid);
+			}
+		}
+	}
+
+	//make the template and send it out
+	tpl_BuildTemplate();
+?>
Index: doc/sql/static-data/data.sql
===================================================================
--- doc/sql/static-data/data.sql	(revision 1892)
+++ doc/sql/static-data/data.sql	(working copy)
@@ -389,26 +389,30 @@
 (2, 0, 'C', 'Nie znaleziona', 'Didn''t find it', 'log/16x16-dnf.png'),
 (3, 0, 'A', 'Komentarz', 'Note', 'log/16x16-note.png'),
 (7, 0, 'C', 'Uczestniczył', 'Attended', 'log/16x16-go.png'),
-(8, 0, 'C', 'Zamierza uczestniczyć', 'Will attended', 'log/16x16-wattend.png');
+(8, 0, 'C', 'Zamierza uczestniczyć', 'Will attended', 'log/16x16-wattend.png'),
+(9, 0, '', 'Solved-pl', 'Solved-en', 'log/16x16-solved.png'),
+(10, 0, '', 'Not solved-pl', 'Not solved-en', 'log/16x16-not_solved.png');
 
 --
 -- Zrzut danych tabeli `log_types_text`
 --
 
-INSERT INTO `log_types_text` (`id`, `log_types_id`, `lang`, `text_combo`, `text_listing`) VALUES
-(1, 1, 'PL', 'Znaleziona', 'skrzynka znaleziona'),
-(2, 2, 'PL', 'Nie znaleziona', 'skrzynka nie znaleziona'),
-(3, 3, 'PL', 'Komentarz', 'komentarz do skrzynki'),
-(4, 4, 'PL', 'Gesperrt', 'hat den Cache gesperrt'),
-(5, 5, 'PL', 'Fregegeben', 'hat den Cache wieder freigegeben'),
-(6, 6, 'PL', 'Entfernt', 'hat den Cache entfernt'),
-(7, 7, 'PL', 'Uczestniczył', 'uczestniczył w wydarzeniu'),
-(8, 8, 'PL', 'Będzie uczestniczył', 'będzie uczestniczył w wydarzeniu'),
-(9, 7, 'EN', 'Attended', 'Attended the event'),
-(10, 8, 'EN', 'Will attend', 'Will attend the event'),
-(11, 1, 'EN', 'Found it', 'Found it'),
-(12, 2, 'EN', 'Didn''t find it', 'Didn''t find it'),
-(13, 3, 'EN', 'Comment', 'Comment');
+INSERT INTO `log_types_text` (`log_types_id`, `lang`, `text_combo`, `text_listing`) VALUES
+(1, 'PL', 'Znaleziona', 'skrzynka znaleziona'),
+(2, 'PL', 'Nie znaleziona', 'skrzynka nie znaleziona'),
+(3, 'PL', 'Komentarz', 'komentarz do skrzynki'),
+(4, 'PL', 'Gesperrt', 'hat den Cache gesperrt'),
+(5, 'PL', 'Fregegeben', 'hat den Cache wieder freigegeben'),
+(6, 'PL', 'Entfernt', 'hat den Cache entfernt'),
+(7, 'PL', 'Uczestniczył', 'uczestniczył w wydarzeniu'),
+(8, 'PL', 'Będzie uczestniczył', 'będzie uczestniczył w wydarzeniu'),
+(7, 'EN', 'Attended', 'Attended the event'),
+(8, 'EN', 'Will attend', 'Will attend the event'),
+(1, 'EN', 'Found it', 'Found it'),
+(2, 'EN', 'Didn''t find it', 'Didn''t find it'),
+(3, 'EN', 'Comment', 'Comment'),
+(9, 'EN', 'Solved it', 'Solved it'),
+(10, 'EN', 'Didn''t solve it', 'Didn''t solve it');
 
 --
 -- Zrzut danych tabeli `nodes`
Index: doc/sql/tables/caches.sql
===================================================================
--- doc/sql/tables/caches.sql	(revision 1892)
+++ doc/sql/tables/caches.sql	(working copy)
@@ -39,6 +39,9 @@
   `votes` int(11) NOT NULL default '0',
   `score` float(2,1) NOT NULL default '0.0',
   `need_npa_recalc` tinyint(1) NOT NULL default '1',
+  `solution` varchar(20) default NULL,
+  `solved` int(11) NOT NULL default '0',
+  `not_solved` int(11) NOT NULL default '0',
   PRIMARY KEY  (`cache_id`),
   UNIQUE KEY `wp_oc` (`wp_oc`),
   KEY `longitude` (`longitude`,`latitude`),
Index: doc/sql/tables/caches_logs.sql
===================================================================
--- doc/sql/tables/caches_logs.sql	(revision 1892)
+++ doc/sql/tables/caches_logs.sql	(working copy)
@@ -17,6 +17,7 @@
   `owner_notified` int(1) NOT NULL default '0',
   `node` tinyint(4) NOT NULL default '0',
   `deleted` tinyint(1) NOT NULL default '0',
+  `hidden` tinyint(1) NOT NULL default '0',
   PRIMARY KEY  (`id`),
   KEY `cache_id` (`cache_id`),
   KEY `user_id` (`user_id`),
Index: editcache.php
===================================================================
--- editcache.php	(revision 1892)
+++ editcache.php	(working copy)
@@ -41,7 +41,7 @@
 		}
 		else
 		{
-			$cache_rs = sql("SELECT `user_id`, `name`, `picturescount`, `mp3count`,`type`, `size`, `date_hidden`, `date_activate`, `longitude`, `latitude`, `country`, `terrain`, `difficulty`, `desc_languages`, `status`, `search_time`, `way_length`, `logpw`, `wp_gc`, `wp_nc` FROM `caches` WHERE `cache_id`='&1'", $cache_id);
+			$cache_rs = sql("SELECT `user_id`, `name`, `picturescount`, `mp3count`,`type`, `size`, `date_hidden`, `date_activate`, `longitude`, `latitude`, `country`, `terrain`, `difficulty`, `desc_languages`, `status`, `search_time`, `way_length`, `logpw`, `wp_gc`, `wp_nc`, `solution` FROM `caches` WHERE `cache_id`='&1'", $cache_id);
 			if (mysql_num_rows($cache_rs) == 1)
 			{
 				$cache_record = sql_fetch_array($cache_rs);
@@ -162,6 +162,7 @@
 					}
 
 					$log_pw = isset($_POST['log_pw']) ? mb_substr($_POST['log_pw'], 0, 20) : $cache_record['logpw'];
+					$solution = isset($_POST['solution']) ? mb_substr($_POST['solution'], 0, 20) : $cache_record['solution'];
 					$wp_gc = isset($_POST['wp_gc']) ? $_POST['wp_gc'] : $cache_record['wp_gc'];
 					$wp_nc = isset($_POST['wp_nc']) ? $_POST['wp_nc'] : $cache_record['wp_nc'];
 
@@ -401,7 +402,7 @@
 							}
 
 							//save to DB
-							sql("UPDATE `caches` SET `last_modified`=NOW(), `name`='&1', `longitude`='&2', `latitude`='&3', `type`='&4', `date_hidden`='&5', `country`='&6', `size`='&7', `difficulty`='&8', `terrain`='&9', `status`='&10', `search_time`='&11', `way_length`='&12', `logpw`='&13', `wp_gc`='&14', `wp_nc`='&15', `date_activate` = $activation_date WHERE `cache_id`='&16'", $cache_name, $cache_lon, $cache_lat, $cache_type, date('Y-m-d', mktime(0, 0, 0, $cache_hidden_month, $cache_hidden_day, $cache_hidden_year)), $cache_country, $sel_size, $cache_difficulty, $cache_terrain, $status, $search_time, $way_length, $log_pw, $wp_gc, $wp_nc, $cache_id);
+							sql("UPDATE `caches` SET `last_modified`=NOW(), `name`='&1', `longitude`='&2', `latitude`='&3', `type`='&4', `date_hidden`='&5', `country`='&6', `size`='&7', `difficulty`='&8', `terrain`='&9', `status`='&10', `search_time`='&11', `way_length`='&12', `logpw`='&13', `wp_gc`='&14', `wp_nc`='&15', `date_activate` = $activation_date, `solution`='&17' WHERE `cache_id`='&16'", $cache_name, $cache_lon, $cache_lat, $cache_type, date('Y-m-d', mktime(0, 0, 0, $cache_hidden_month, $cache_hidden_day, $cache_hidden_year)), $cache_country, $sel_size, $cache_difficulty, $cache_terrain, $status, $search_time, $way_length, $log_pw, $wp_gc, $wp_nc, $cache_id, $solution);
 
 							// delete old cache-attributes
 							sql("DELETE FROM `caches_attributes` WHERE `cache_id`='&1'", $cache_id);
@@ -854,6 +855,7 @@
 
 					tpl_set_var('way_length', $way_length);
 					tpl_set_var('log_pw', htmlspecialchars($log_pw, ENT_COMPAT, 'UTF-8'));
+					tpl_set_var('solution', htmlspecialchars($solution, ENT_COMPAT, 'UTF-8'));
 					tpl_set_var('wp_gc', htmlspecialchars($wp_gc, ENT_COMPAT, 'UTF-8'));
 					tpl_set_var('wp_nc', htmlspecialchars($wp_nc, ENT_COMPAT, 'UTF-8'));
 
Index: editlog.php
===================================================================
--- editlog.php	(revision 1892)
+++ editlog.php	(working copy)
@@ -20,6 +20,19 @@
 	require_once('./lib/common.inc.php');
 	require($stylepath.'/smilies.inc.php');
 
+	function isSolvedLogType($log_type)
+	{
+		return $log_type == 9 || $log_type == 10;
+	}
+	
+	function enableRating($log_record)
+	{
+		if ($log_record['cachetype'] == 6)
+			return false;
+
+		return !isSolvedLogType($log_record['logtype']);
+	}
+
 	//Preprocessing
 	if ($error == false)
 	{
@@ -135,14 +148,15 @@
 						$rating_msg = mb_ereg_replace('{curr}', $user_tops, $rating_msg);
 					}
 
-// sp2ong 28.I.2010 recommendation all caches except events
-				if ( $log_record['cachetype'] != 6 ) {
-					tpl_set_var('rating_message', mb_ereg_replace('{rating_msg}', $rating_msg, $rating_tpl));
+					if (enableRating($log_record))
+					{
+						tpl_set_var('rating_message', mb_ereg_replace('{rating_msg}', $rating_msg, $rating_tpl));
+					}
+					else
+					{
+						tpl_set_var('rating_message', ""); 
+					}
 
-				} else {
-				tpl_set_var('rating_message', ""); 			
-				}
-				
 					if (isset($_POST['descMode']))
 					{
 						$descMode = $_POST['descMode']+0;
@@ -276,154 +290,162 @@
 					}
 
 					//store?
-					if (isset($_POST['submitform']) && $date_not_ok == false && $logtype_not_ok == false && $pw_not_ok == false)
+					if (isset($_POST['submitform']))
 					{
-						//store changed data
-						sql("UPDATE `cache_logs` SET `type`='&1',
-						                             `date`='&2',
-						                             `text`='&3',
-						                             `text_html`='&4',
-						                             `text_htmledit`='&5',
-						                             `last_modified`=NOW()
-						                       WHERE `id`='&6'",
-						                             $log_type,
-						                             date('Y-m-d H:i:s', mktime($log_date_hour, $log_date_min, 0, $log_date_month, $log_date_day, $log_date_year)),
-						                             tidy_html_description((($descMode != 1) ? $log_text : nl2br($log_text))),
-						                             (($descMode != 1) ? 1 : 0),
-						                             (($descMode == 3) ? 1 : 0),
-						                             $log_id);
-
-						//update user-stat if type changed
-						if ($log_record['logtype'] != $log_type)
+						if (isSolvedLogType($log_type))
 						{
-							$user_rs = sql("SELECT `founds_count`, `notfounds_count`, `log_notes_count` FROM `user` WHERE `user_id`='&1'", $log_record['user_id']);
-							$user_record = sql_fetch_array($user_rs);
-							mysql_free_result($user_rs);
+							sql("UPDATE `cache_logs` SET `text`='&1', `text_html`='&2', `text_htmledit`='&3', `last_modified`=NOW() WHERE `id`='&4'",
+									tidy_html_description((($descMode != 1) ? $log_text : nl2br($log_text))), (($descMode != 1) ? 1 : 0), (($descMode == 3) ? 1 : 0), $log_id);
+						}
+						else if ($date_not_ok == false && $logtype_not_ok == false && $pw_not_ok == false)
+						{
+							//store changed data
+							sql("UPDATE `cache_logs` SET `type`='&1',
+														`date`='&2',
+														`text`='&3',
+														`text_html`='&4',
+														`text_htmledit`='&5',
+														`last_modified`=NOW()
+												WHERE `id`='&6'",
+														$log_type,
+														date('Y-m-d H:i:s', mktime($log_date_hour, $log_date_min, 0, $log_date_month, $log_date_day, $log_date_year)),
+														tidy_html_description((($descMode != 1) ? $log_text : nl2br($log_text))),
+														(($descMode != 1) ? 1 : 0),
+														(($descMode == 3) ? 1 : 0),
+														$log_id);
 
-							if ($log_record['logtype'] == 1 || $log_record['logtype'] == 7)
+							//update user-stat if type changed
+							if ($log_record['logtype'] != $log_type)
 							{
-								$user_record['founds_count']--;
-								
-								// recalc scores for this cache
-								sql("DELETE FROM `scores` WHERE `user_id` = '&1' AND `cache_id` = '&2'", $log_record['user_id'], $log_record['cache_id']);
-								$sql = "SELECT count(*) FROM scores WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
-								$liczba = mysql_result(mysql_query($sql),0);
-								$sql = "SELECT SUM(score) FROM scores WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
-								$suma = @mysql_result(@mysql_query($sql),0)+0;
+								$user_rs = sql("SELECT `founds_count`, `notfounds_count`, `log_notes_count` FROM `user` WHERE `user_id`='&1'", $log_record['user_id']);
+								$user_record = sql_fetch_array($user_rs);
+								mysql_free_result($user_rs);
 
-								// obliczenie nowej sredniej
-								if( $liczba != 0)
+								if ($log_record['logtype'] == 1 || $log_record['logtype'] == 7)
 								{
-									$srednia = $suma / $liczba;
+									$user_record['founds_count']--;
+									
+									// recalc scores for this cache
+									sql("DELETE FROM `scores` WHERE `user_id` = '&1' AND `cache_id` = '&2'", $log_record['user_id'], $log_record['cache_id']);
+									$sql = "SELECT count(*) FROM scores WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
+									$liczba = mysql_result(mysql_query($sql),0);
+									$sql = "SELECT SUM(score) FROM scores WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
+									$suma = @mysql_result(@mysql_query($sql),0)+0;
+
+									// obliczenie nowej sredniej
+									if( $liczba != 0)
+									{
+										$srednia = $suma / $liczba;
+									}
+									else 
+									{
+										$srednia = 0;
+									}
+									
+									$sql = "UPDATE caches SET votes='".sql_escape($liczba)."', score='".sql_escape($srednia)."' WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
+									mysql_query($sql);
 								}
-								else 
+								elseif ($log_record['logtype'] == 2)
 								{
-									$srednia = 0;
+									$user_record['notfounds_count']--;
 								}
-								
-								$sql = "UPDATE caches SET votes='".sql_escape($liczba)."', score='".sql_escape($srednia)."' WHERE cache_id='".sql_escape($log_record['cache_id'])."'";
-								mysql_query($sql);
-							}
-							elseif ($log_record['logtype'] == 2)
-							{
-								$user_record['notfounds_count']--;
-							}
-							elseif ($log_record['logtype'] == 3)
-							{
-								$user_record['log_notes_count']--;
-							}
+								elseif ($log_record['logtype'] == 3)
+								{
+									$user_record['log_notes_count']--;
+								}
 
-							// falls eines der felder NULL
-							$user_record['founds_count'] = $user_record['founds_count']+0;
-							$user_record['notfounds_count'] = $user_record['notfounds_count']+0;
-							$user_record['log_notes_count'] = $user_record['log_notes_count']+0;
+								// falls eines der felder NULL
+								$user_record['founds_count'] = $user_record['founds_count']+0;
+								$user_record['notfounds_count'] = $user_record['notfounds_count']+0;
+								$user_record['log_notes_count'] = $user_record['log_notes_count']+0;
 
-							if ($log_type == 1 || $log_type == 7)
-							{
-								$user_record['founds_count']++;
+								if ($log_type == 1 || $log_type == 7)
+								{
+									$user_record['founds_count']++;
+								}
+								elseif ($log_type == 2)
+								{
+									$user_record['notfounds_count']++;
+									if( $res['founds'] <= 1)
+										$top_cache = 0;
+								}
+								elseif ($log_type == 3)
+								{
+									$user_record['log_notes_count']++;
+									if( $res['founds'] <= 1)
+										$top_cache = 0;
+								}
+
+								sql("UPDATE `user` SET `founds_count`='&1', `notfounds_count`='&2', `log_notes_count`='&3' WHERE `user_id`='&4'", $user_record['founds_count'], $user_record['notfounds_count'], $user_record['log_notes_count'], $log_record['user_id']);
+								unset($user_record);
+
+								//call eventhandler
+								require_once($rootpath . 'lib/eventhandler.inc.php');
+								event_change_log_type($log_record['cache_id'], $log_record['user_id']+0);
 							}
-							elseif ($log_type == 2)
-							{
-								$user_record['notfounds_count']++;
-								if( $res['founds'] <= 1)
-									$top_cache = 0;
-							}
-							elseif ($log_type == 3)
-							{
-								$user_record['log_notes_count']++;
-								if( $res['founds'] <= 1)
-									$top_cache = 0;
-							}
 
-							sql("UPDATE `user` SET `founds_count`='&1', `notfounds_count`='&2', `log_notes_count`='&3' WHERE `user_id`='&4'", $user_record['founds_count'], $user_record['notfounds_count'], $user_record['log_notes_count'], $log_record['user_id']);
-							unset($user_record);
+							//update cache-stat if type or log_date changed
+							$cache_rs = sql("SELECT `founds`, `notfounds`, `notes` FROM `caches` WHERE `cache_id`='&1'", $log_record['cache_id']);
+							$cache_record = sql_fetch_array($cache_rs);
+							mysql_free_result($cache_rs);
 
-							//call eventhandler
-							require_once($rootpath . 'lib/eventhandler.inc.php');
-							event_change_log_type($log_record['cache_id'], $log_record['user_id']+0);
-						}
+							if ($log_record['logtype'] != $log_type)
+							{
+								if ($log_record['logtype'] == 1 || $log_record['logtype'] == 7)
+								{
+									$cache_record['founds']--;
+								}
+								elseif ($log_record['logtype'] == 2 || $log_record['logtype'] == 8)
+								{
+									$cache_record['notfounds']--;
+								}
+								elseif ($log_record['logtype'] == 3)
+								{
+									$cache_record['notes']--;
+								}
 
-						//update cache-stat if type or log_date changed
-						$cache_rs = sql("SELECT `founds`, `notfounds`, `notes` FROM `caches` WHERE `cache_id`='&1'", $log_record['cache_id']);
-						$cache_record = sql_fetch_array($cache_rs);
-						mysql_free_result($cache_rs);
+								// falls eines der felder NULL
+								$cache_record['founds'] = $cache_record['founds']+0;
+								$cache_record['notfounds'] = $cache_record['notfounds']+0;
+								$cache_record['notes'] = $cache_record['notes']+0;
 
-						if ($log_record['logtype'] != $log_type)
-						{
-							if ($log_record['logtype'] == 1 || $log_record['logtype'] == 7)
-							{
-								$cache_record['founds']--;
+								if ($log_type == 1 || $log_type == 7)
+								{
+									$cache_record['founds']++;
+								}
+								elseif ($log_type == 2 || $log_type == 8)
+								{
+									$cache_record['notfounds']++;
+								}
+								elseif ($log_type == 3)
+								{
+									$cache_record['notes']++;
+								}
 							}
-							elseif ($log_record['logtype'] == 2 || $log_record['logtype'] == 8)
-							{
-								$cache_record['notfounds']--;
-							}
-							elseif ($log_record['logtype'] == 3)
-							{
-								$cache_record['notes']--;
-							}
 
-							// falls eines der felder NULL
-							$cache_record['founds'] = $cache_record['founds']+0;
-							$cache_record['notfounds'] = $cache_record['notfounds']+0;
-							$cache_record['notes'] = $cache_record['notes']+0;
+							// update top-list
+							if ($top_cache == 1)
+								sql("INSERT IGNORE INTO `cache_rating` (`user_id`, `cache_id`) VALUES('&1', '&2')", $log_record['user_id'], $log_record['cache_id']);
+							else
+								sql("DELETE FROM `cache_rating` WHERE `user_id`='&1' AND `cache_id`='&2'", $log_record['user_id'], $log_record['cache_id']);
 
-							if ($log_type == 1 || $log_type == 7)
+							//Update last found
+							$lastfound_rs = sql("SELECT MAX(`cache_logs`.`date`) AS `date` FROM `cache_logs` WHERE ((`cache_logs`.`type`=1) AND (`cache_logs`.`cache_id`='&1') AND deleted=&2)", $log_record['cache_id'], 0);
+							$lastfound_record = sql_fetch_array($lastfound_rs);
+
+							if ($lastfound_record['date'] === NULL)
 							{
-								$cache_record['founds']++;
+								$lastfound = 'NULL';
 							}
-							elseif ($log_type == 2 || $log_type == 8)
+							else
 							{
-								$cache_record['notfounds']++;
+								$lastfound = $lastfound_record['date'];
 							}
-							elseif ($log_type == 3)
-							{
-								$cache_record['notes']++;
-							}
-						}
 
-						// update top-list
-						if ($top_cache == 1)
-							sql("INSERT IGNORE INTO `cache_rating` (`user_id`, `cache_id`) VALUES('&1', '&2')", $log_record['user_id'], $log_record['cache_id']);
-						else
-							sql("DELETE FROM `cache_rating` WHERE `user_id`='&1' AND `cache_id`='&2'", $log_record['user_id'], $log_record['cache_id']);
-
-						//Update last found
-						$lastfound_rs = sql("SELECT MAX(`cache_logs`.`date`) AS `date` FROM `cache_logs` WHERE ((`cache_logs`.`type`=1) AND (`cache_logs`.`cache_id`='&1') AND deleted=&2)", $log_record['cache_id'], 0);
-						$lastfound_record = sql_fetch_array($lastfound_rs);
-
-						if ($lastfound_record['date'] === NULL)
-						{
-							$lastfound = 'NULL';
+							sql("UPDATE `caches` SET `last_found`='&1', `founds`='&2', `notfounds`='&3', `notes`='&4' WHERE `cache_id`='&5'", $lastfound, $cache_record['founds'], $cache_record['notfounds'], $cache_record['notes'], $log_record['cache_id']);
+							unset($cache_record);
 						}
-						else
-						{
-							$lastfound = $lastfound_record['date'];
-						}
 
-						sql("UPDATE `caches` SET `last_found`='&1', `founds`='&2', `notfounds`='&3', `notes`='&4' WHERE `cache_id`='&5'", $lastfound, $cache_record['founds'], $cache_record['notfounds'], $cache_record['notes'], $log_record['cache_id']);
-						unset($cache_record);
-
 						//display cache page
 						tpl_redirect('viewcache.php?cacheid=' . urlencode($log_record['cache_id']));
 						exit;
@@ -560,6 +582,19 @@
 						}
 					}
 					tpl_set_var('smilies', $smilies);
+
+					if (isSolvedLogType($log_type))
+					{
+						tpl_set_var('hide_type_and_date_start', '<!--');
+						tpl_set_var('hide_type_and_date_end', '-->');
+						tpl_set_var('hidden_log_type', '<input class="input40" type="text" name="logtype" value="'.$logtype.'"/>');
+					}
+					else
+					{
+						tpl_set_var('hide_type_and_date_start', '');
+						tpl_set_var('hide_type_and_date_end', '');
+						tpl_set_var('hidden_log_type', '');
+					}
 				}
 				else
 				{
Index: lib/caches.inc.php
===================================================================
--- lib/caches.inc.php	(revision 1892)
+++ lib/caches.inc.php	(working copy)
@@ -52,6 +52,9 @@
 	$resp = sql("SELECT * FROM log_types ORDER BY id ASC");
 	while($row = sql_fetch_assoc($resp))
 	{
+		if ($row['id'] == 9 || $row['id'] == 10)
+			continue;
+
 		$log_types[] = $row;
 	}
 	return $log_types;
Index: lib/t_solved.php
===================================================================
--- lib/t_solved.php	(revision 0)
+++ lib/t_solved.php	(revision 0)
@@ -0,0 +1,75 @@
+<?php
+	global $lang, $rootpath;
+
+	if (!isset($rootpath)) $rootpath = './';
+
+	//include template handling
+	require_once($rootpath . 'lib/common.inc.php');
+
+  setlocale(LC_TIME, 'sv_SE.UTF-8');
+$rsU = sql('SELECT COUNT(*) `count` FROM (SELECT COUNT(cache_logs.user_id) FROM `cache_logs` WHERE type=9 AND `deleted`=0 AND `hidden`=0 GROUP BY `user_id`) `users_with_founds`');
+$fC = sql('SELECT COUNT(*) `count` FROM `cache_logs` WHERE type=9 AND `deleted`=0 AND `hidden`=0');
+  $rsUs = mysql_fetch_array($rsU);
+    $fCt = mysql_fetch_array($fC);
+
+	echo '<center><table width="97%" border="0"><tr><td align="center"><center><b>'.tr('ranking_by_number_of_solved').'</b><br />'.tr('number_of_users_who_has_solved').':';
+	echo $rsUs[count]; 
+	echo ' .::. '.tr('number_of_solved').':';
+	echo $fCt[count]; 
+	echo '</center></td></tr>';
+
+	echo '</table>';
+	echo '<table border="1" bgcolor="white" width="97%" style="font-size:11px; line-height:1.6em;">' . "\n";
+
+$a = "SELECT COUNT(*) count, username, stat_ban, user.user_id FROM cache_logs, user ".
+     "WHERE `cache_logs`.`deleted`=0 AND `cache_logs`.`hidden`=0 AND cache_logs.user_id=user.user_id AND cache_logs.type=9 ".
+     "GROUP BY user.user_id ".
+     "ORDER BY 1 DESC, user.username ASC";
+
+echo "<br />";
+
+$r=mysql_query($a) or die(mysql_error());
+echo    '<tr class="bgcolor2">'.
+        '<td align="center">&nbsp;&nbsp;<b>'.tr('ranking').'</b>&nbsp;&nbsp;</td>'.
+        '<td align="center"><b>'.tr('number_of_solved').'</b></td>'.
+        '<td align="center">&nbsp;&nbsp;<b>'.tr('username').'</b>&nbsp;&nbsp;</td></tr><tr><td>';
+
+$l2=""; // number of users within the same rank
+$position=1; // position ex aequo; incremented by number of users in each rank
+
+while ($line=mysql_fetch_array($r))
+{
+	$color = "black";
+	$banned = "";
+	if( $usr['admin'] || $line['stat_ban'] == 0)
+	{
+		if( $line['stat_ban'] )
+		{
+			$color = "gray";
+			$banned = " (BAN)";
+		}
+    $l1=$line[count];
+    if ($l2!=$l1)
+    {
+      echo '</td></tr>';
+      echo '<tr class="bgcolor2">'.
+         '<td align="right">&nbsp;&nbsp;<b>'.$position.'</b>&nbsp;&nbsp;</td>'.
+         '<td align="right">&nbsp;&nbsp;<b>'.$l1.'</b>&nbsp;&nbsp;</td>'.
+         '<td>';
+      $l2=$l1;
+    }
+    else 
+    {
+      echo ', ';
+    }
+
+    echo '<a style="color:'.$color.'" href="viewprofile.php?userid='.$line[user_id].'">'.htmlspecialchars($line[username]).$banned.'</a>';
+    $position++;
+	}
+}
+
+// end table
+//echo "</td></tr>";
+echo "</table>\n";
+
+?>
Index: my_logs.php
===================================================================
--- my_logs.php	(revision 1892)
+++ my_logs.php	(working copy)
@@ -57,7 +57,8 @@
 		
 	$rs = sql("SELECT count(id) FROM cache_logs, caches WHERE `cache_logs`.`cache_id`=`caches`.`cache_id`
 				AND `cache_logs`.`deleted`=0 
-			    AND `caches`.`status` != 4
+				AND `cache_logs`.`hidden`=0 
+				AND `caches`.`status` != 4
 				AND `caches`.`status` != 5 
 				AND `caches`.`status` != 6
 				AND `cache_logs`.`user_id`='" . sql_escape($_REQUEST['userid']) . "'");
@@ -96,7 +97,8 @@
 			FROM `cache_logs`, `caches`
 			WHERE `cache_logs`.`cache_id`=`caches`.`cache_id`
 				AND `cache_logs`.`deleted`=0 
-			  AND `caches`.`status` != 4
+				AND `cache_logs`.`hidden`=0 
+				AND `caches`.`status` != 4
 				AND `caches`.`status` != 5 
 				AND `caches`.`status` != 6
 				AND `cache_logs`.`user_id`='" . sql_escape($_REQUEST['userid']) . "'
Index: mycaches_logs.php
===================================================================
--- mycaches_logs.php	(revision 1892)
+++ mycaches_logs.php	(working copy)
@@ -57,7 +57,8 @@
 	$rs = sql("SELECT count(id) FROM cache_logs, caches 
 			WHERE `cache_logs`.`cache_id`=`caches`.`cache_id`
 				AND `cache_logs`.`deleted`=0 
-			  AND `caches`.`status` != 4
+				AND `cache_logs`.`hidden`=0 
+				AND `caches`.`status` != 4
 				AND `caches`.`status` != 5 
 				AND `caches`.`status` != 6
 				AND `caches`.`user_id`='" . sql_escape($_REQUEST['userid']) . "'");
@@ -97,7 +98,8 @@
 			FROM `cache_logs`, `caches`
 			WHERE `cache_logs`.`cache_id`=`caches`.`cache_id`
 				AND `cache_logs`.`deleted`=0 
-			  AND `caches`.`status` != 4
+				AND `cache_logs`.`hidden`=0 
+				AND `caches`.`status` != 4
 				AND `caches`.`status` != 5 
 				AND `caches`.`status` != 6
 				AND `caches`.`user_id`='" . sql_escape($_REQUEST['userid']) . "'
Index: myhome.php
===================================================================
--- myhome.php	(revision 1892)
+++ myhome.php	(working copy)
@@ -229,6 +229,7 @@
 					WHERE `caches`.`user_id`='&1'
 					AND `cache_logs`.`cache_id`=`caches`.`cache_id` 
 					AND `cache_logs`.`deleted`=0 
+					AND `cache_logs`.`hidden`=0 
 					AND `user`.`user_id`=`cache_logs`.`user_id`
 					AND `log_types`.`id`=`cache_logs`.`type`
 					AND `log_types_text`.`log_types_id`=`log_types`.`id` AND `log_types_text`.`lang`='&2'
Index: newcache.php
===================================================================
--- newcache.php	(revision 1892)
+++ newcache.php	(working copy)
@@ -358,6 +358,9 @@
 				$log_pw = isset($_POST['log_pw']) ? mb_substr($_POST['log_pw'], 0, 20) : '';
 				tpl_set_var('log_pw', htmlspecialchars($log_pw, ENT_COMPAT, 'UTF-8'));
 
+				$solution = isset($_POST['solution']) ? mb_substr($_POST['solution'], 0, 20) : '';
+				tpl_set_var('solution', htmlspecialchars($solution, ENT_COMPAT, 'UTF-8'));
+				
 				// gc- and nc-waypoints
 				$wp_gc = isset($_POST['wp_gc']) ? $_POST['wp_gc'] : '';
 				tpl_set_var('wp_gc', htmlspecialchars($wp_gc, ENT_COMPAT, 'UTF-8'));
@@ -957,10 +960,11 @@
 													`way_length`,
 													`wp_gc`,
 													`wp_nc`,
-													`node`
+													`node`,
+													`solution`
 												) VALUES (
 													'', '&1', '&2', '&3', '&4', NOW(), NOW(), '&5', '&6', '&7', '&8', $activation_date, '0', '0', '0', NULL ,
-													'&9', '&10', '&11', '&12', '&13', '&14', '&15', '&16', '&17', '&18')",
+													'&9', '&10', '&11', '&12', '&13', '&14', '&15', '&16', '&17', '&18', '&19')",
 												$usr['userid'],
 												$name,
 												$longitude,
@@ -978,7 +982,8 @@
 												$way_length,
 												$wp_gc,
 												$wp_nc,
-												$oc_nodeid);
+												$oc_nodeid,
+												$solution);
 						$cache_id = mysql_insert_id($dblink);
 
 						// waypoint erstellen
Index: newcaches.php
===================================================================
--- newcaches.php	(revision 1892)
+++ newcaches.php	(working copy)
@@ -77,7 +77,7 @@
 							LEFT JOIN	gk_item_waypoint ON gk_item_waypoint.wp = caches.wp_oc
 							LEFT JOIN	gk_item ON gk_item.id = gk_item_waypoint.id AND
 							gk_item.stateid<>1 AND gk_item.stateid<>4 AND gk_item.typeid<>2 AND gk_item.stateid !=5				
-			WHERE cache_logs.deleted=0 AND cache_logs.cache_id=&1
+			WHERE cache_logs.deleted=0 AND cache_logs.hidden=0 AND cache_logs.cache_id=&1
 			 GROUP BY cache_logs.id ORDER BY cache_logs.date_created DESC LIMIT 1",$r['cacheid']);
 
 			if (mysql_num_rows($rs_log) != 0)
Index: newlogs.php
===================================================================
--- newlogs.php	(revision 1892)
+++ newlogs.php	(working copy)
@@ -43,7 +43,7 @@
 	$LOGS_PER_PAGE = 50;
 	$PAGES_LISTED = 10;
 		
-	$rs = sql("SELECT count(id) FROM cache_logs WHERE deleted=0");
+	$rs = sql("SELECT count(id) FROM cache_logs WHERE deleted=0 and hidden=0");
 	$total_logs = mysql_result($rs,0);
 	mysql_free_result($rs);
 	
@@ -83,7 +83,8 @@
 			FROM `cache_logs`, `caches`
 			WHERE `cache_logs`.`cache_id`=`caches`.`cache_id`
 				AND `cache_logs`.`deleted`=0 
-			  AND `caches`.`status` != 4
+				AND `cache_logs`.`hidden`=0 
+				AND `caches`.`status` != 4
 				AND `caches`.`status` != 5 
 				AND `caches`.`status` != 6
 			ORDER BY  `cache_logs`.`date_created` DESC
Index: search.php
===================================================================
--- search.php	(revision 1892)
+++ search.php	(working copy)
@@ -713,6 +713,7 @@
 					$sql_where[] = '`caches`.`cache_id`=`cache_logs`.`cache_id`';
 					$sql_where[] = '`cache_logs`.`user_id`=\'' . sql_escape($finder_id) . '\'';
 					$sql_where[] = '`cache_logs`.`deleted`=0';
+					$sql_where[] = '`cache_logs`.`hidden`=0';
 					if( $options['logtype'] == "" )
 						$sql_where[] = '(`cache_logs`.`type`=1 OR `cache_logs`.`type`=7)'; // found und attended
 					else
Index: tpl/stdstyle/articles/s_solved.tpl.php
===================================================================
--- tpl/stdstyle/articles/s_solved.tpl.php	(revision 0)
+++ tpl/stdstyle/articles/s_solved.tpl.php	(revision 0)
@@ -0,0 +1,11 @@
+<table class="content" width="97%">
+	<tr><td class="content2-pagetitle"><img src="tpl/stdstyle/images/blue/stat1.png" class="icon32" alt="{{stats}}" title="{{stats}}" align="middle" /><font size="4">  <b>{{stats}}</b></font></td></tr>
+	<tr><td class="spacer"></td></tr>
+</table>
+
+<table class="table" width="760" style="line-height: 1.6em; font-size: 10px;">
+<tr>
+<td><?php include ("t_solved.php");?>
+</td></tr>
+</table>
+
Index: tpl/stdstyle/articles/stat.tpl.php
===================================================================
--- tpl/stdstyle/articles/stat.tpl.php	(revision 1892)
+++ tpl/stdstyle/articles/stat.tpl.php	(working copy)
@@ -4,6 +4,7 @@
 <img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s2">{{ranking_by_number_of_finds}}</a><br />
 <img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s6">{{ranking_by_number_of_recommnedations}}</a><br />
 <img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s3">{{user_ranking_by_number_of_finds_of_their_caches}}</a><br />
+<img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s_solved">{{ranking_by_number_of_solved}}</a><br />
 <hr align ="left" style="border: 0; width: 500px;color: #000000; background-color: #000000;height: 1px;"/>
 <img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s4">{{cache_ranking_by_number_of_finds}}</a><br />
 <img src="tpl/stdstyle/images/free_icons/chart_bar.png" class="icon16" alt="" title="" align="middle" />&nbsp;<a class="links" href="articles.php?page=s11a">Ranking skrzynek według liczby odkryć w województwie</a><br />
Index: tpl/stdstyle/checksolution.inc.php
===================================================================
--- tpl/stdstyle/checksolution.inc.php	(revision 0)
+++ tpl/stdstyle/checksolution.inc.php	(revision 0)
@@ -0,0 +1,33 @@
+<?php
+/***************************************************************************
+											./tpl/stdstyle/editlog.inc.php
+															-------------------
+		begin                : Mon July 5 2004
+		copyright            : (C) 2004 The OpenCaching Group
+		forum contact at     : http://www.opencaching.com/phpBB2
+
+	***************************************************************************/
+
+/***************************************************************************
+	*
+	*   This program is free software; you can redistribute it and/or modify
+	*   it under the terms of the GNU General Public License as published by
+	*   the Free Software Foundation; either version 2 of the License, or
+	*   (at your option) any later version.
+	*
+	***************************************************************************/
+
+/****************************************************************************
+
+   Unicode Reminder メモ
+
+	 language vars
+
+ ****************************************************************************/
+
+	$submit = 'Submit';
+
+ 	$error_wrong_node = "This entry in the log has been provided by another Opencaching server and you can edit it only on that server.";
+
+	$smiley_link = '<a href="javascript:insertSmiley(\'{smiley_text}\')">{smiley_image}</a>';
+?>
Index: tpl/stdstyle/checksolution.tpl.php
===================================================================
--- tpl/stdstyle/checksolution.tpl.php	(revision 0)
+++ tpl/stdstyle/checksolution.tpl.php	(revision 0)
@@ -0,0 +1,354 @@
+<?php
+/***************************************************************************
+	*                                         				                                
+	*   This program is free software; you can redistribute it and/or modify  	
+	*   it under the terms of the GNU General Public License as published by  
+	*   the Free Software Foundation; either version 2 of the License, or	    	
+	*   (at your option) any later version.
+	*   
+	*  UTF-8 ąść
+	***************************************************************************/
+?>
+<script type="text/javascript">
+<!--
+function insertSmiley(parSmiley) {
+  var myText = document.checksolution.logtext;
+  myText.focus();
+  /* fuer IE */
+  if(typeof document.selection != 'undefined') {
+    var range = document.selection.createRange();
+    var selText = range.text;
+    range.text = parSmiley + selText;
+  }
+  /* fuer Firefox/Mozilla-Browser */
+  else if(typeof myText.selectionStart != 'undefined')
+  {
+    var start = myText.selectionStart;
+    var end = myText.selectionEnd;
+    var selText = myText.value.substring(start, end);
+    myText.value = myText.value.substr(0, start) + parSmiley + selText + myText.value.substr(end);
+    /* Cursorposition hinter Smiley setzen */
+    myText.selectionStart = start + parSmiley.length;
+    myText.selectionEnd = start + parSmiley.length;
+  }
+  /* fuer die anderen Browser */
+  else
+  {
+    alert(navigator.appName + ': Setting smilies is not supported');
+  }
+}
+
+function toogleLayer( whichLayer, val )
+{
+	var elem, vis;
+
+	if( document. getElementById )
+		elem = document.getElementById(whichLayer);
+	else if( document.all )
+		elem = document.all[whichLayer];
+	else if( document.layers )
+		elem = document.layers[whichLayer];
+	vis = elem.style;
+	
+	if(val != '')
+	{
+	if (document.checksolution.logtype.value == "1" || document.checksolution.logtype.value == "7") 
+		vis.display = 'block';
+	else
+		vis.display = 'none';
+	}
+	else
+		vis.display = val;
+}
+//-->
+</script>
+<form action="checksolution.php" method="post" enctype="application/x-www-form-urlencoded" name="checksolution" dir="ltr">
+<input type="hidden" name="logid" value="{logid}"/>
+<input type="hidden" name="version2" value="1"/>
+<input id="descMode" type="hidden" name="descMode" value="1" />
+<table class="content">
+	<tr><td class="content2-pagetitle" colspan="2"><img src="tpl/stdstyle/images/blue/logs.png" class="icon32" alt="" title="" align="middle" /> <b>{{check_solution_log}} <a href="viewcache.php?cacheid={cacheid}">{cachename}</a></b></td></tr>
+</table>
+<table class="content" style="font-size: 12px; line-height: 1.6em;">
+	<tr><td class="spacer" colspan="2"></td></tr>
+	<tr>
+		<td>{solution_status}</td>
+	</tr>
+	<tr><td class="spacer" colspan="2"></td></tr>
+</table>
+<table class="content" style="font-size: 12px; line-height: 1.6em;">
+	<tr>
+		<td colspan="2"><br /><img src="tpl/stdstyle/images/free_icons/page_edit.png" class="icon16" alt="" title="" align="middle" />&nbsp;<strong>{{comments_log}}:</td>
+	</tr>
+	<tr>
+		<td colspan="2">
+			<div class="menuBar">
+				<span id="descText" class="buttonNormal" onclick="btnSelect(1)" onmouseover="btnMouseOver(1)" onmouseout="btnMouseOut(1)">Text</span>
+				<span class="buttonSplitter">|</span>
+				<span id="descHtml" class="buttonNormal" onclick="btnSelect(2)" onmouseover="btnMouseOver(2)" onmouseout="btnMouseOut(2)">&lt;html&gt;</span>
+				<span class="buttonSplitter">|</span>
+				<span id="descHtmlEdit" class="buttonNormal" onclick="btnSelect(3)" onmouseover="btnMouseOver(3)" onmouseout="btnMouseOut(3)">Editor</span>
+			</div>
+		</td>
+	</tr>
+	<tr>
+		<td>
+			<span id="scriptwarning" class="errormsg">Javascript is turned off by your browser. You may only enter plain text. To switch and use the HTML code editor you have to enable Javascript.</span>
+		</td>
+	</tr>
+	<tr>
+		<td>
+			<textarea name="logtext" id="logtext" cols="68" rows="25" >{logtext}</textarea>
+    </td>
+	</tr>
+	<tr>
+		<td colspan="2">
+			{smilies}
+		</td>
+	</tr>
+	<tr>
+		<td class="header-small" colspan="2">
+			<input type="submit" name="cancelform" value="{{cancel}}" style="width:120px"/>&nbsp;&nbsp;
+			<input type="submit" name="submitform" value="{{submit}}" style="width:120px"/>
+		</td>
+	</tr>
+</table>
+</form>
+<script language="javascript" type="text/javascript">
+<!--
+	/*
+		1 = Text
+		2 = HTML
+		3 = HTML-Editor
+	*/
+	var use_tinymce = 0;
+	var descMode = {descMode};
+	document.getElementById("scriptwarning").firstChild.nodeValue = "";
+
+	// descMode auf 1 oder 2 stellen ... wenn Editor erfolgreich geladen wieder auf 3 Stellen
+	if (descMode == 3)
+	{
+		if (document.getElementById("logtext").value == '')
+			descMode = 1;
+		else
+			descMode = 2;
+	}
+
+	document.getElementById("descMode").value = descMode;
+	mnuSetElementsNormal();
+
+	function postInit()
+	{
+		descMode = 3;
+		use_tinymce = 1;
+		document.getElementById("descMode").value = descMode;
+		mnuSetElementsNormal();
+	}
+
+	function toggleEditor(id) {
+		if (!tinyMCE.getInstanceById(id))
+			tinyMCE.execCommand('mceAddControl', false, id);
+		else
+			tinyMCE.execCommand('mceRemoveControl', false, id);
+	}
+
+
+	function SwitchToTextDesc(oldMode)
+	{
+		document.getElementById("descMode").value = 1;
+
+		if(use_tinymce)
+			toggleEditor("logtext");
+		use_tinymce = 0;
+		// convert HTML to text
+		var desc = document.getElementById("logtext").value;
+
+		desc = html_entity_decode(desc, ['ENT_NOQUOTES']);
+
+		document.getElementById("logtext").value = desc;
+	}
+
+	function SwitchToHtmlDesc(oldMode)
+	{
+		document.getElementById("descMode").value = 2;
+
+		if(use_tinymce)
+			toggleEditor("logtext");
+		use_tinymce = 0;
+
+		// convert text to HTML
+		var desc = document.getElementById("logtext").value;
+
+		if(oldMode != 3)
+			desc = htmlspecialchars(desc, ['ENT_NOQUOTES']);
+
+		document.getElementById("logtext").value = desc;
+	}
+
+	function SwitchToHtmlEditDesc(oldMode)
+	{
+		document.getElementById("descMode").value = 3;
+		use_tinymce = 1;
+
+		if(oldMode == 2) {
+			var desc = document.getElementById("logtext").value;
+			desc = html_entity_decode(desc, ['ENT_NOQUOTES']);
+			document.getElementById("logtext").value = desc;
+		}
+
+		toggleEditor("logtext");
+	}
+
+	function mnuSelectElement(e)
+	{
+		e.backgroundColor = '#D4D5D8';
+		e.borderColor = '#6779AA';
+		e.borderWidth = '1px';
+		e.borderStyle = 'solid';
+	}
+
+	function mnuNormalElement(e)
+	{
+		e.backgroundColor = '#F0F0EE';
+		e.borderColor = '#F0F0EE';
+		e.borderWidth = '1px';
+		e.borderStyle = 'solid';
+	}
+
+	function mnuHoverElement(e)
+	{
+		e.backgroundColor = '#B6BDD2';
+		e.borderColor = '#0A246A';
+		e.borderWidth = '1px';
+		e.borderStyle = 'solid';
+	}
+
+	function mnuUnhoverElement(e)
+	{
+		mnuSetElementsNormal();
+	}
+
+	function mnuSetElementsNormal()
+	{
+		var descText = document.getElementById("descText").style;
+		var descHtml = document.getElementById("descHtml").style;
+		var descHtmlEdit = document.getElementById("descHtmlEdit").style;
+
+		switch (descMode)
+		{
+			case 1:
+				mnuSelectElement(descText);
+				mnuNormalElement(descHtml);
+				mnuNormalElement(descHtmlEdit);
+
+				break;
+			case 2:
+				mnuNormalElement(descText);
+				mnuSelectElement(descHtml);
+				mnuNormalElement(descHtmlEdit);
+
+				break;
+			case 3:
+				mnuNormalElement(descText);
+				mnuNormalElement(descHtml);
+				mnuSelectElement(descHtmlEdit);
+
+				break;
+		}
+	}
+
+	function btnSelect(mode)
+	{
+		var descText = document.getElementById("descText").style;
+		var descHtml = document.getElementById("descHtml").style;
+		var descHtmlEdit = document.getElementById("descHtmlEdit").style;
+
+		var oldMode = descMode;
+		descMode = mode;
+		mnuSetElementsNormal();
+
+		if(oldMode == descMode)
+	{
+			// convert text to HTML
+			var desc = document.getElementById("logtext").value;
+
+			if ((desc.indexOf('&amp;') == -1) &&
+			    (desc.indexOf('&quot;') == -1) &&
+			    (desc.indexOf('&lt;') == -1) &&
+			    (desc.indexOf('&gt;') == -1) &&
+			    (desc.indexOf('<p>') == -1) &&
+			    (desc.indexOf('<i>') == -1) &&
+			    (desc.indexOf('<strong>') == -1) &&
+			    (desc.indexOf('<br />') == -1))
+			{
+				desc = desc.replace(/&/g, "&amp;");
+				desc = desc.replace(/"/g, "&quot;");
+				desc = desc.replace(/</g, "&lt;");
+				desc = desc.replace(/>/g, "&gt;");
+				desc = desc.replace(/\r\n/g, "\<br />");
+				desc = desc.replace(/\n/g, "<br />");
+				desc = desc.replace(/<br \/>/g, "<br />\n");
+			}
+
+			document.getElementById("logtext").value = desc;
+		}
+
+
+		switch (mode)
+		{
+			case 1:
+				SwitchToTextDesc(oldMode);
+				break;
+			case 2:
+				SwitchToHtmlDesc(oldMode);
+				break;
+			case 3:
+
+				SwitchToHtmlEditDesc(oldMode);
+				break;
+		}
+	}
+
+	function btnMouseOver(id)
+	{
+		var descText = document.getElementById("descText").style;
+		var descHtml = document.getElementById("descHtml").style;
+		var descHtmlEdit = document.getElementById("descHtmlEdit").style;
+
+		switch (id)
+		{
+			case 1:
+				mnuHoverElement(descText);
+				break;
+			case 2:
+				mnuHoverElement(descHtml);
+				break;
+			case 3:
+				mnuHoverElement(descHtmlEdit);
+				break;
+		}
+	}
+
+	function btnMouseOut(id)
+	{
+		var descText = document.getElementById("descText").style;
+		var descHtml = document.getElementById("descHtml").style;
+		var descHtmlEdit = document.getElementById("descHtmlEdit").style;
+
+		switch (id)
+		{
+			case 1:
+				mnuUnhoverElement(descText);
+				break;
+			case 2:
+				mnuUnhoverElement(descHtml);
+				break;
+			case 3:
+				mnuUnhoverElement(descHtmlEdit);
+				break;
+		}
+	}
+
+
+//-->
+</script>
Index: tpl/stdstyle/css/style_screen.css
===================================================================
--- tpl/stdstyle/css/style_screen.css	(revision 1892)
+++ tpl/stdstyle/css/style_screen.css	(working copy)
@@ -517,6 +517,7 @@
 textarea.logs {		width:600px; height:300px; }	/* only logeditor */
 
 .show_deleted { text-decoration: line-through; }
+.show_hidden { text-decoration: overline; }
 
 div#new-caches-area { width: 500px; }
 div#new-events-area { width: 500px; }
Index: tpl/stdstyle/editcache.tpl.php
===================================================================
--- tpl/stdstyle/editcache.tpl.php	(revision 1892)
+++ tpl/stdstyle/editcache.tpl.php	(working copy)
@@ -41,6 +41,10 @@
 			document.editcache_form.size.options[document.editcache_form.size.options.length - 1 ] = null;
 		document.editcache_form.size.disabled = false;
   }
+
+  document.getElementById("solutioncheck").style.visibility  = document.editcache_form.type.value == "7" ? "visible" : "hidden";
+  document.getElementById("solutioncheck2").style.visibility  = document.editcache_form.type.value == "7" ? "visible" : "hidden";
+
   return false;
 }
 
@@ -296,6 +300,11 @@
 		<td><input class="input100" type="text" name="log_pw" value="{log_pw}" maxlength="20"/> ({{no_password_label}})</td>
 	</tr>
 	<tr><td colspan="2"><div class="notice buffer" style="width:500px;height:24px;">{{please_read}}</div></td></tr>
+	<tr id="solutioncheck" style="visibility: hidden;">
+		<td><p class="content-title-noshade">{{solution_password}}:</p></td>
+		<td><input class="input100" type="text" name="solution" value="{solution}" maxlength="20"/> ({{solution_password_label}})</td>
+	</tr>
+	<tr id="solutioncheck2" style="visibility: hidden;"><td colspan="2"><div class="notice buffer" style="width:650px;height:35px; line-height: 1.4em;">{{solution_password_please_read}}</div></td></tr>
 	<tr><td colspan="2"><div class="errormsg"><br />{{creating_cache}}<br /></div></td></tr>
 
 	<tr><td class="buffer" colspan="2"></td></tr>
Index: tpl/stdstyle/editlog.tpl.php
===================================================================
--- tpl/stdstyle/editlog.tpl.php	(revision 1892)
+++ tpl/stdstyle/editlog.tpl.php	(working copy)
@@ -91,11 +91,12 @@
 	<tr><td class="content2-pagetitle" colspan="2"><img src="tpl/stdstyle/images/blue/logs.png" class="icon32" alt="" title="edit log Cache" align="middle" /> <b>Edycja logu dla skrzynki <a href="viewcache.php?cacheid={cacheid}">{cachename}</a></b></td></tr>
 </table>
 <table class="content" style="font-size: 12px; line-height: 1.6em;">
+{hide_type_and_date_start}
 	<tr><td class="spacer" colspan="2"></td></tr>
 	<tr>
 		<td width="180px"><img src="tpl/stdstyle/images/free_icons/page_go.png" class="icon16" alt="" title="" align="middle" />&nbsp;<strong>{{type_of_log}}:</strong></td>
 		<td align="left">
-			<!--<select name="logtype" onChange="return _chkFound()">-->
+			{hidden_log_type}
 			<select onload="javascript:toogleLayer('ocena');" name="logtype" onchange="javascript:toogleLayer('ocena');">
 				{logtypeoptions}
 			</select>
@@ -114,6 +115,7 @@
 			<br />{date_message}
 		</td>
 	</tr>
+{hide_type_and_date_end}
 	<tr><td class="spacer" colspan="2"></td></tr>
 	{rating_message}
 </table>
Index: tpl/stdstyle/images/log/16x16-not_solved.png
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream

Property changes on: tpl\stdstyle\images\log\16x16-not_solved.png
___________________________________________________________________
Added: svn:mime-type
   + application/octet-stream

Index: tpl/stdstyle/images/log/16x16-solved.png
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream

Property changes on: tpl\stdstyle\images\log\16x16-solved.png
___________________________________________________________________
Added: svn:mime-type
   + application/octet-stream

Index: tpl/stdstyle/newcache.tpl.php
===================================================================
--- tpl/stdstyle/newcache.tpl.php	(revision 1892)
+++ tpl/stdstyle/newcache.tpl.php	(working copy)
@@ -88,6 +88,9 @@
 			document.newcacheform.size.options[document.newcacheform.size.options.length - 2 ] = null;
 		document.newcacheform.size.disabled = false;
   }
+
+  document.getElementById("solutioncheck").style.visibility  = document.newcacheform.type.value == "7" ? "visible" : "hidden";
+
   return false;
 }
 
@@ -422,6 +425,15 @@
 		</td>
 	</tr>
 	<tr><td colspan="2"><div class="notice buffer" style="width:500px;height:24px;">{{please_read}}</div></td></tr>
+	<tr>
+	<td colspan="2"><div id="solutioncheck" style="visibility: hidden;">
+		<fieldset style="border: 1px solid black; width: 80%; height: 32%; background-color: #FFFFFF;">
+		<legend>&nbsp; <strong>{{solution_password}}</strong> &nbsp;</legend>
+		<input class="input100" type="text" name="solution" value="{solution}" maxlength="20"/> ({{solution_password_label}})
+		</fieldset>
+		<div class="notice buffer" style="width:650px;height:35px; line-height: 1.4em;">{{solution_password_please_read}}</div>
+		</div></td>
+	</tr>
 	<tr><td colspan="2"><div class="errormsg"><br />{{creating_cache}}<br /><br /></div></td></tr>
 	<tr>
 		<td colspan="2">
Index: tpl/stdstyle/viewcache.inc.php
===================================================================
--- tpl/stdstyle/viewcache.inc.php	(revision 1892)
+++ tpl/stdstyle/viewcache.inc.php	(working copy)
@@ -107,6 +107,8 @@
 
  $cache_found_text = "x ".tr('found');
  $cache_notfound_text = "x ".tr('not_found');
+ $cache_solved_text = "x ".tr('solved');
+ $cache_not_solved_text = "x ".tr('not_solved');
 
  $recommend_link = '&nbsp;&nbsp;<a href="recommendations.php?cacheid={cacheid}"/>('.tr('show_recommended').')</a>';
  $rating_stat_show_singular = '<img src="images/rating-star.png" alt="{{recomendation}}" /> {ratings} '.tr('recommendation').'<br />';
@@ -114,6 +116,8 @@
 
 $found_icon = '<img src="tpl/stdstyle/images/log/16x16-found.png" class="icon16" alt="{{found}}" />';
 $notfound_icon = '<img src="tpl/stdstyle/images/log/16x16-dnf.png" class="icon16" alt="{{not_found}}" />';
+$solved_icon = '<img src="tpl/stdstyle/images/log/16x16-solved.png" class="icon16" alt="{{solved}}" />';
+$not_solved_icon = '<img src="tpl/stdstyle/images/log/16x16-not_solved.png" class="icon16" alt="{{not_solved}}" />';
 $note_icon = '<img src="tpl/stdstyle/images/log/16x16-note.png" class="icon16" alt="{{comment}}" />';
 $notes_icon = '<img src="tpl/stdstyle/images/free_icons/note_edit.png" class="icon16" alt="" />';
 $vote_icon = '<img src="tpl/stdstyle/images/free_icons/medal_bronze_1.png" class="icon16" alt="" />';
Index: tpl/stdstyle/viewcache.tpl.php
===================================================================
--- tpl/stdstyle/viewcache.tpl.php	(revision 1892)
+++ tpl/stdstyle/viewcache.tpl.php	(working copy)
@@ -109,8 +109,12 @@
 							{found_icon} {founds} {found_text}<br />
 							{notfound_icon} {notfounds} {notfound_text}<br />
 							{note_icon} {notes} {{comments}}<br />
-							{notes_icon} {cache_notes} {cachenotes_link}<br />
-							{gk_icon} <a class="links" href="http://geokrety.org/szukaj.php?lang=pl_PL.UTF-8&wpt={oc_waypoint}" target="_blank">{{history_gk}}</a><br />
+							{notes_icon} {cache_notes} {cachenotes_link}<br />
+							{gk_icon} <a class="links" href="http://geokrety.org/szukaj.php?lang=pl_PL.UTF-8&wpt={oc_waypoint}" target="_blank">{{history_gk}}</a><br />
+							{solution_start}
+							{solved_icon} {solved} {solved_text}<br />
+							{not_solved_icon} {not_solved} {not_solved_text}<br />
+							{solution_end}
 							{watch_icon} {watcher} {{watchers}}<br />
 							{visit_icon} {visits} {{visitors}}<br />
 							{vote_icon} {votes_count} x {{scored}}<br />
@@ -327,7 +331,36 @@
 ";
 	 ?>
 				</div>
+{solution_start}
+<script type='text/javascript'>
+function notEmpty(elem, helperMsg)
+{
+	if (elem.value.length == 0)
+	{
+		alert(helperMsg);
+		elem.focus();
+		return false;
+	}
+	return true;
+}
+</script>
+			<div class="content2-container bg-blue02">
+				<p class="content-title-noshade-size1">
+					<img src="tpl/stdstyle/images/blue/decrypt.png" class="icon32" alt="" /><img src="tpl/stdstyle/images/blue/password.png" class="icon32" alt="" />&nbsp;{{check_solution}}
+				</p>
 			</div>
+			<div class="content2-container">
+				<form name="solution_check" action="checksolution.php" method="get" onSubmit="return notEmpty(document.getElementById('solution'), '{{solution_not_empty}}')">
+					<input type="hidden" name="cacheid" value="{cacheid}"/>
+					<p>{{solution_check_intro}}</p>
+					<input class="input100" type="text" name="solution" id="solution" maxlength="100" value="" {solution_disabled} />
+					<button type="submit" style="font-size:12px; width:120px" {solution_disabled}>{{submit_check_solution}}</button>
+					<br /><br />
+					<div class="notice buffer" id="viewcache-solutioninfo" style="height:32px;">{solution_check_info}</div>
+				</form>
+			</div>
+{solution_end}
+			</div>
 <!-- Text container -->
 			<div class="content2-container bg-blue02">
 				<p class="content-title-noshade-size1">
@@ -337,6 +370,10 @@
 					{found_icon} {founds}x
 					{notfound_icon} {notfounds}x
 					{note_icon} {notes}x
+					{solution_start}
+					{solved_icon} {solved_logs}x
+					{not_solved_icon} {not_solved_logs}x
+					{solution_end}
 					&nbsp;&nbsp;
 					{viewlogs}
 					&nbsp;
Index: tpl/stdstyle/viewcache_log.tpl.php
===================================================================
--- tpl/stdstyle/viewcache_log.tpl.php	(revision 1892)
+++ tpl/stdstyle/viewcache_log.tpl.php	(working copy)
@@ -1,4 +1,4 @@
-	<div class="content-txtbox-noshade line-box {show_deleted}">
+	<div class="content-txtbox-noshade line-box {show_deleted}{show_hidden}">
 	<p class="content-title-noshade-size1">{logimage} {date} {ratingimage} <a href="viewprofile.php?userid={userid}">{username}</a> {type} {logfunctions}</p> 
 	<div class="viewcache_log-content">
 	{logtext}
Index: tpl/stdstyle/viewlogs.tpl.php
===================================================================
--- tpl/stdstyle/viewlogs.tpl.php	(revision 1892)
+++ tpl/stdstyle/viewlogs.tpl.php	(working copy)
@@ -33,6 +33,10 @@
 			{found_icon} {founds}x 
 			{notfound_icon} {notfounds}x 
 			{note_icon} {notes}x
+			{solution_start}
+			{solved_icon} {solved_logs}x
+			{not_solved_icon} {not_solved_logs}x
+			{solution_end}
 				</p>
 			</div>
 <div class="content2-container" id="viewcache-logs">
Index: util.sec/watchlist/runwatch.php
===================================================================
--- util.sec/watchlist/runwatch.php	(revision 1892)
+++ util.sec/watchlist/runwatch.php	(working copy)
@@ -52,7 +52,7 @@
 /* end db connect */
   
 /* begin owner notifies */
-  $rsNewLogs = sql("SELECT cache_logs.id log_id, caches.user_id user_id FROM cache_logs, caches WHERE cache_logs.deleted=0 AND cache_logs.cache_id=caches.cache_id AND cache_logs.owner_notified=0");
+  $rsNewLogs = sql("SELECT cache_logs.id log_id, caches.user_id user_id FROM cache_logs, caches WHERE cache_logs.deleted=0 AND cache_logs.hidden=0 AND cache_logs.cache_id=caches.cache_id AND cache_logs.owner_notified=0");
   for ($i = 0; $i < mysql_num_rows($rsNewLogs); $i++)
   {
 		$rNewLog = sql_fetch_array($rsNewLogs);
Index: viewcache.php
===================================================================
--- viewcache.php	(revision 1892)
+++ viewcache.php	(working copy)
@@ -32,6 +32,87 @@
 		return -1;
 	}
 
+	function hasStatistics($cache_record)
+	{
+		return ($cache_record['founds'] + $cache_record['notfounds'] + $cache_record['notes'] + $cache_record['solved'] + $cache_record['not_solved']) > 0;
+	}
+
+	function getNumberOfLogs($cache_id)
+	{
+		$logs = 0;
+		$rs = sql("SELECT COUNT(*) logs FROM `cache_logs` WHERE `cache_id` = &1 AND `deleted` = 0 AND `hidden` = 0", $cache_id);
+
+		if ($r = sql_fetch_assoc($rs))
+		{
+			$logs = $r['logs'];
+		}
+
+		mysql_free_result($rs);
+
+		return $logs;
+	}
+
+	function hasSolution($cache_record)
+	{
+		return $cache_record['solution'] != NULL && $cache_record['solution'] != '';
+	}
+
+	function isCurUserOwner($cache_record)
+	{
+		global $usr;
+
+		return $cache_record['user_id'] == $usr['userid'];
+	}
+
+	function isSolutionEnabled($cache_record)
+	{
+		global $usr;
+
+		return $usr == true && !isCurUserOwner($cache_record);
+	}
+
+	function isSolvedByUser($cache_id)
+	{
+		global $usr;
+		
+		$solved = 0;
+		$rs = sql("SELECT COUNT(*) solved FROM `cache_logs` WHERE `cache_id` = &1 AND `user_id` = &2 AND `type` = 9 AND `deleted` = 0 AND `hidden` = 0", $cache_id, $usr['userid']);
+
+		if ($r = sql_fetch_assoc($rs))
+		{
+			$solved = $r['solved'];
+		}
+
+		mysql_free_result($rs);
+
+		return $solved > 0;
+	}
+
+	function getNumberOfLogsByType($cache_id, $log_type)
+	{
+		$logs = 0;
+		$rs = sql("SELECT COUNT(*) logs FROM `cache_logs` WHERE `cache_id` = &1 AND `type` = &2 AND `deleted` = 0 AND `hidden` = 0", $cache_id, $log_type);
+
+		if ($r = sql_fetch_assoc($rs))
+		{
+			$logs = $r['logs'];
+		}
+
+		mysql_free_result($rs);
+
+		return $logs;
+	}
+
+	function getNumberOfSolvedLogs($cache_id)
+	{
+		return getNumberOfLogsByType($cache_id, 9);
+	}
+
+	function getNumberOfNotSolvedLogs($cache_id)
+	{
+		return getNumberOfLogsByType($cache_id, 10);
+	}
+	
 	//Preprocessing
 	if ($error == false)
 	{
@@ -131,6 +212,9 @@
 								`caches`.`topratings` `topratings`,
 								`caches`.`ignorer_count` `ignorer_count`,
 								`caches`.`votes` `votes_count`,
+								`caches`.`solution` `solution`,
+								`caches`.`solved` `solved`,
+								`caches`.`not_solved` `not_solved`,
 								`cache_type`.`icon_large` `icon_large`,
 			                  `user`.`username` `username`,
 							  `countries`.`&1` AS `country_name`,
@@ -153,7 +237,7 @@
 				$cache_record = sql_fetch_array($rs);
 			}
 			mysql_free_result($rs);
-			if( $cache_record['user_id'] == $usr['userid'] || $usr['admin'])
+			if (isCurUserOwner($cache_record) || $usr['admin'])
 			{
 				$show_edit = true;
 			}
@@ -325,15 +409,18 @@
 
 
 			if ($cache_record['type'] == 6 ) 
-			{$cache_stats='';
-			} else { 
-			if (($cache_record['founds'] + $cache_record['notfounds'] + $cache_record['notes']) != 0) 
 			{
-			$cache_stats = "<a class =\"links2\" href=\"javascript:void(0)\" onmouseover=\"Tip('" .tr('show_statictics_cache'). "', BALLOON, true, ABOVE, false, OFFSETX, -17, PADDING, 8, WIDTH, -240)\" onmouseout=\"UnTip()\" onclick=\"javascript:window.open('cache_stats.php?cacheid=".$cache_record['cache_id']."&amp;popup=y','Cache_Statistics','width=500,height=750,resizable=yes,scrollbars=1')\"><img src=\"tpl/stdstyle/images/blue/stat1.png\" alt=\"Statystyka skrzynki\" title=\"Statystyka skrzynki\" /></a>";
-			} else {
+				$cache_stats='';
+			}
+			else if (hasStatistics($cache_record)) 
+			{
+				$cache_stats = "<a class =\"links2\" href=\"javascript:void(0)\" onmouseover=\"Tip('" .tr('show_statistics_cache'). "', BALLOON, true, ABOVE, false, OFFSETX, -17, PADDING, 8, WIDTH, -240)\" onmouseout=\"UnTip()\" onclick=\"javascript:window.open('cache_stats.php?cacheid=".$cache_record['cache_id']."&amp;popup=y','Cache_Statistics','width=500,height=750,resizable=yes,scrollbars=1')\"><img src=\"tpl/stdstyle/images/blue/stat1.png\" alt=\"Statystyka skrzynki\" title=\"Statystyka skrzynki\" /></a>";
+			}
+			else
+			{
 			$cache_stats="<a class =\"links2\" href=\"javascript:void(0)\" onmouseover=\"Tip('" .tr('not_stat_cache'). "', BALLOON, true, ABOVE, false, OFFSETX, -17, PADDING, 8, WIDTH, -240)\" onmouseout=\"UnTip()\"><img src=\"tpl/stdstyle/images/blue/stat1.png\" alt=\"\" title=\"\" /></a>";
 					}
-			}			
+
 			tpl_set_var('cache_stats', $cache_stats);
 			tpl_set_var('googlemap_key', $googlemap_key);
 			tpl_set_var('map_msg', $map_msg);
@@ -527,18 +614,21 @@
 			tpl_set_var('founds', htmlspecialchars($cache_record['founds'], ENT_COMPAT, 'UTF-8'));
 			tpl_set_var('notfounds', htmlspecialchars($cache_record['notfounds'], ENT_COMPAT, 'UTF-8'));
 			tpl_set_var('notes', htmlspecialchars($cache_record['notes'], ENT_COMPAT, 'UTF-8'));
-			tpl_set_var('total_number_of_logs', htmlspecialchars($cache_record['notes'] + $cache_record['notfounds'] + $cache_record['founds'], ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('solved', htmlspecialchars($cache_record['solved'], ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('not_solved', htmlspecialchars($cache_record['not_solved'], ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('solved_logs', htmlspecialchars(getNumberOfSolvedLogs($cache_id), ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('not_solved_logs', htmlspecialchars(getNumberOfNotSolvedLogs($cache_id), ENT_COMPAT, 'UTF-8'));
 
-			// number of cache notes
-			$crs = sql("SELECT COUNT(*) as `count` FROM `cache_notes` WHERE (`cache_id`='&1' AND user_id='&2')", $cache_id, $usr['userid']);
-			if (mysql_num_rows($crs) == 0){
-				tpl_set_var('cache_notes', '0');
-				} else {
-				$cachenotes_record = mysql_fetch_array($crs);
-				tpl_set_var('cache_notes', $cachenotes_record['count']);
-			}
-			tpl_set_var('cachenotes_link', '<a class="links" href="mycache_notes.php?cacheid='.$cache_id.'">'.tr('cachenotes').'</a>');
-			mysql_free_result($crs);
+			// number of cache notes
+			$crs = sql("SELECT COUNT(*) as `count` FROM `cache_notes` WHERE (`cache_id`='&1' AND user_id='&2')", $cache_id, $usr['userid']);
+			if (mysql_num_rows($crs) == 0){
+				tpl_set_var('cache_notes', '0');
+				} else {
+				$cachenotes_record = mysql_fetch_array($crs);
+				tpl_set_var('cache_notes', $cachenotes_record['count']);
+			}
+			tpl_set_var('cachenotes_link', '<a class="links" href="mycache_notes.php?cacheid='.$cache_id.'">'.tr('cachenotes').'</a>');
+			mysql_free_result($crs);
 			tpl_set_var('watcher', $cache_record['watcher'] + 0);
 			tpl_set_var('ignorer_count', $cache_record['ignorer_count'] + 0);
 			tpl_set_var('votes_count', $cache_record['votes_count'] + 0);
@@ -566,10 +656,14 @@
 			{
 				tpl_set_var('found_icon', $found_icon);
 				tpl_set_var('notfound_icon', $notfound_icon);
+				tpl_set_var('solved_icon', $solved_icon);
+				tpl_set_var('not_solved_icon', $not_solved_icon);
 
 				tpl_set_var('event_attendance_list', '');
 				tpl_set_var('found_text', $cache_found_text);
 				tpl_set_var('notfound_text', $cache_notfound_text);
+				tpl_set_var('solved_text', $cache_solved_text);
+				tpl_set_var('not_solved_text', $cache_not_solved_text);
 			}
 
 			// number of visits
@@ -583,7 +677,7 @@
 			}
 
 
-			if (($cache_record['founds'] + $cache_record['notfounds'] + $cache_record['notes']) > $logs_to_display)
+			if (getNumberOfLogs($cache_id) > $logs_to_display)
 			{
 				tpl_set_var('viewlogs_last', mb_ereg_replace('{cacheid_urlencode}', htmlspecialchars(urlencode($cache_id), ENT_COMPAT, 'UTF-8'), $viewlogs_last));
 				tpl_set_var('viewlogs', mb_ereg_replace('{cacheid_urlencode}', htmlspecialchars(urlencode($cache_id), ENT_COMPAT, 'UTF-8'), $viewlogs));
@@ -921,10 +1015,10 @@
 			// prepare the last n logs - show logs marked as deleted if admin
 			//
 			$show_deleted_logs = "";
-			$show_deleted_logs2 = " AND `cache_logs`.`deleted` = 0 ";
+			$show_deleted_logs2 = " AND `cache_logs`.`deleted` = 0 AND `cache_logs`.`hidden` = 0 ";
 			if( $usr['admin'] )
 			{
-				$show_deleted_logs = "`cache_logs`.`deleted` `deleted`,";
+				$show_deleted_logs = "`cache_logs`.`deleted` `deleted`, `cache_logs`.`hidden` `hidden`,";
 				$show_deleted_logs2 = "";
 			}
 			
@@ -960,6 +1054,12 @@
 				{
 					$show_deleted = "show_deleted";
 				}
+
+				$show_hidden = "";
+				if( isset( $record['hidden'] ) && $record['hidden'] )
+				{
+					$show_hidden = "show_hidden";
+				}
 				
 				$tmplog = read_file($stylepath . '/viewcache_log.tpl.php');
 
@@ -1033,6 +1133,7 @@
 					$tmplog = mb_ereg_replace('{logfunctions}', '', $tmplog);
 
 				$tmplog = mb_ereg_replace('{show_deleted}', $show_deleted, $tmplog);
+				$tmplog = mb_ereg_replace('{show_hidden}', $show_hidden, $tmplog);
 				$tmplog = mb_ereg_replace('{username}', $tmplog_username, $tmplog);
 				$tmplog = mb_ereg_replace('{userid}', $record['userid'], $tmplog);
 				$tmplog = mb_ereg_replace('{date}', $tmplog_date, $tmplog);
@@ -1251,9 +1352,28 @@
 				tpl_set_var('cache_attributes', '');
 				tpl_set_var('password_req', '');
 			}
+
+			tpl_set_var('solution_disabled', 'disabled="disabled"');
+			tpl_set_var('solution_check_info', tr('solution_check_info'));
+
+			if (hasSolution($cache_record))
+			{
+				tpl_set_var('solution_start', '');
+				tpl_set_var('solution_end', '');
+
+				if (isSolvedByUser($cache_id))
+					tpl_set_var('solution_check_info', tr('is_solved_by_user'));
+				else if (isSolutionEnabled($cache_record))
+					tpl_set_var('solution_disabled', '');
 		}
 		else
 		{
+				tpl_set_var('solution_start', '<!--');
+				tpl_set_var('solution_end', '-->');
+			}
+		}
+		else
+		{
 			//display search page
 			$tplname = 'viewcache_error';
 		}
Index: viewlogs.php
===================================================================
--- viewlogs.php	(revision 1892)
+++ viewlogs.php	(working copy)
@@ -32,6 +32,36 @@
   //prepare the templates and include all neccessary
 	require_once('./lib/common.inc.php');
 	
+	function hasSolution($cache_record)
+	{
+		return $cache_record['solution'] != NULL && $cache_record['solution'] != '';
+	}
+
+	function getNumberOfLogsByType($cache_id, $log_type)
+	{
+		$logs = 0;
+		$rs = sql("SELECT COUNT(*) logs FROM `cache_logs` WHERE `cache_id` = &1 AND `type` = &2 AND `deleted` = 0 AND `hidden` = 0", $cache_id, $log_type);
+
+		if ($r = sql_fetch_assoc($rs))
+		{
+			$logs = $r['logs'];
+		}
+
+		mysql_free_result($rs);
+
+		return $logs;
+	}
+
+	function getNumberOfSolvedLogs($cache_id)
+	{
+		return getNumberOfLogsByType($cache_id, 9);
+	}
+
+	function getNumberOfNotSolvedLogs($cache_id)
+	{
+		return getNumberOfLogsByType($cache_id, 10);
+	}
+
 	//Preprocessing
 	if ($error == false)
 	{
@@ -65,7 +95,7 @@
 		if ($cache_id != 0)
 		{
 			//get cache record
-			$rs = sql("SELECT `user_id`, `name`, `founds`, `notfounds`, `notes`, `status`, `type` FROM `caches` WHERE `caches`.`cache_id`='&1'", $cache_id);
+			$rs = sql("SELECT `user_id`, `name`, `founds`, `notfounds`, `notes`, `status`, `type`, `solution` FROM `caches` WHERE `caches`.`cache_id`='&1'", $cache_id);
 
 			if (mysql_num_rows($rs) == 0)
 			{
@@ -101,21 +131,35 @@
 			{
 				tpl_set_var('found_icon', $found_icon);
 				tpl_set_var('notfound_icon', $notfound_icon);
+				tpl_set_var('solved_icon', $solved_icon);
+				tpl_set_var('not_solved_icon', $not_solved_icon);
 			}
 		    tpl_set_var('note_icon', $note_icon);
 
 			tpl_set_var('founds', htmlspecialchars($cache_record['founds'], ENT_COMPAT, 'UTF-8'));
 			tpl_set_var('notfounds', htmlspecialchars($cache_record['notfounds'], ENT_COMPAT, 'UTF-8'));
 			tpl_set_var('notes', htmlspecialchars($cache_record['notes'], ENT_COMPAT, 'UTF-8'));
-			tpl_set_var('total_number_of_logs', htmlspecialchars($cache_record['notes'] + $cache_record['notfounds'] + $cache_record['founds'], ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('solved_logs', htmlspecialchars(getNumberOfSolvedLogs($cache_id), ENT_COMPAT, 'UTF-8'));
+			tpl_set_var('not_solved_logs', htmlspecialchars(getNumberOfNotSolvedLogs($cache_id), ENT_COMPAT, 'UTF-8'));
+
+			if (hasSolution($cache_record))
+			{
+				tpl_set_var('solution_start', '');
+				tpl_set_var('solution_end', '');
+			}
+			else
+			{
+				tpl_set_var('solution_start', '<!--');
+				tpl_set_var('solution_end', '-->');
+			}
 			
 			// prepare the logs - show logs marked as deleted if admin
 			//
 			$show_deleted_logs = "";
-			$show_deleted_logs2 = " AND `cache_logs`.`deleted` = 0 ";
+			$show_deleted_logs2 = " AND `cache_logs`.`deleted` = 0  AND `cache_logs`.`hidden` = 0 ";
 			if( $usr['admin'] )
 			{
-				$show_deleted_logs = "`cache_logs`.`deleted` `deleted`,";
+				$show_deleted_logs = "`cache_logs`.`deleted` `deleted`, `cache_logs`.`hidden` `hidden`,";
 				$show_deleted_logs2 = "";
 			}
 			
@@ -151,6 +195,11 @@
 				{
 					$show_deleted = "show_deleted";
 				}
+				$show_hidden = "";
+				if( isset( $record['hidden'] ) && $record['hidden'] )
+				{
+					$show_hidden = "show_hidden";
+				}
 				$tmplog = read_file($stylepath . '/viewcache_log.tpl.php');
 
 				$tmplog_username = htmlspecialchars($record['username'], ENT_COMPAT, 'UTF-8');
@@ -164,6 +213,7 @@
 				$tmplog_text = tidy_html_description($tmplog_text);
 
 				$tmplog = mb_ereg_replace('{show_deleted}', $show_deleted, $tmplog);
+				$tmplog = mb_ereg_replace('{show_hidden}', $show_hidden, $tmplog);
 				$tmplog = mb_ereg_replace('{username}', $tmplog_username, $tmplog);
 				$tmplog = mb_ereg_replace('{userid}', $record['userid'], $tmplog);
 				$tmplog = mb_ereg_replace('{date}', $tmplog_date, $tmplog);
Index: viewprofile.php
===================================================================
--- viewprofile.php	(revision 1892)
+++ viewprofile.php	(working copy)
@@ -12,7 +12,32 @@
 //prepare the templates and include all neccessary
 	require_once('./lib/common.inc.php');
 		global $stat_menu; $lang;
-	
+
+	function getNumberOfLogsByType($user_id, $log_type)
+	{
+		$sql = "SELECT COUNT(*) founds_count FROM cache_logs WHERE user_id=$user_id AND type=$log_type AND deleted=0 AND hidden=0";
+
+		if ($odp = mysql_query($sql))
+			return mysql_result($odp, 0);
+
+		return 0;
+	}
+
+	function getNumberOfFounds($user_id)
+	{
+		return getNumberOfLogsByType($user_id, 1);
+	}
+
+	function getNumberOfEvents($user_id)
+	{
+		return getNumberOfLogsByType($user_id, 7);
+	}
+
+	function getNumberOfSolved($user_id)
+	{
+		return getNumberOfLogsByType($user_id, 9);
+	}
+
 	//Preprocessing
 	if ($error == false)
 	{
@@ -95,19 +120,14 @@
 			$seek = mysql_num_rows($rs_seek);
 
 		$content .= '<br /><p>&nbsp;</p><div class="content2-container bg-blue02"><p class="content-title-noshade-size1">&nbsp;<img src="tpl/stdstyle/images/blue/cache-open.png" class="icon32" alt="Caches Find" title="Caches Find" />&nbsp;&nbsp;&nbsp;'.tr(stat_number_found).'</p></div><br />';
-		if ($seek == 0) {
+		if ($user_record['founds_count'] == 0)
+		{
 			$content .= '<br /><p> <b>'.tr('not_found_caches').'</b></p>';
 						  }
 						  else 
 						  { 
-			$sql = "SELECT COUNT(*) events_count 
-							FROM cache_logs 
-							WHERE user_id=$user_id AND type=7 AND deleted=0";
-			
-			if( $odp = mysql_query($sql) )
-				$events_count = mysql_result($odp,0);
-			else 
-				$events_count = 0;
+			$events_count = getNumberOfEvents($user_id);
+			$solved_count = getNumberOfSolved($user_id);
 			$days_since_first_find = @mysql_result(@mysql_query("SELECT datediff(now(), date) as old FROM cache_logs WHERE deleted=0 AND user_id = $user_id AND type=1 ORDER BY date LIMIT 1"),0);					   
 			$rsfc2=sql("SELECT cache_logs.cache_id cache_id,  DATE_FORMAT(cache_logs.date,'%d-%m-%Y') data, caches.wp_oc cache_wp FROM cache_logs, caches WHERE caches.cache_id=cache_logs.cache_id AND cache_logs.type='1' AND cache_logs.user_id=&1 AND cache_logs.deleted='0' ORDER BY cache_logs.date DESC LIMIT 1",$user_id);
 			$rfc2 = mysql_fetch_array($rsfc2);
@@ -115,13 +135,7 @@
 			$rc = sql_fetch_array($rsc);
 			$rsncd= sql ("SELECT COUNT(*) FROM cache_logs WHERE type=1 AND cache_logs.deleted='0' AND user_id=&1 GROUP BY YEAR(`date`), MONTH(`date`), DAY(`date`)",$user_id);
 			$num_rows = mysql_num_rows($rsncd);
-			$sql = "SELECT COUNT(*) founds_count 
-					FROM cache_logs 
-					WHERE user_id=$user_id AND type=1 AND deleted=0";
-			if( $odp = mysql_query($sql) )
-			$found = mysql_result($odp,0);
-			else 
-			$found = 0;
+			$found = getNumberOfFounds($user_id);
 			if( $ddays['diff'] == 0 )
 			{
 				$aver1 = 0;
@@ -157,6 +171,12 @@
 				else
 			{ $content .= '&nbsp;&nbsp;&nbsp;<img src="tpl/stdstyle/images/blue/arrow.png" alt="" /> [<a class="links" href="search.php?showresult=1&amp;expert=0&amp;f_inactive=0&amp;output=HTML&amp;sort=bycreated&amp;finderid=' . $user_id . '&amp;searchbyfinder=&amp;logtype=7&amp;f_ignored=0&amp;f_userfound=0&amp;f_userowner=0">'.tr('show').'</a>]</p>';}
 
+			$content .= '<p><span class="content-title-noshade txt-blue08">'.tr('total_solved').':</span> <strong>' . $solved_count . '</strong>';
+			if ($events_count == 0) 
+				{$content .= '</p>';}
+				else
+			{ $content .= '&nbsp;&nbsp;&nbsp;<img src="tpl/stdstyle/images/blue/arrow.png" alt="" /> [<a class="links" href="search.php?showresult=1&amp;expert=0&amp;f_inactive=0&amp;output=HTML&amp;sort=bycreated&amp;finderid=' . $user_id . '&amp;searchbyfinder=&amp;logtype=9&amp;f_ignored=0&amp;f_userfound=0&amp;f_userowner=0">'.tr('show').'</a>]</p>';}
+
 			$recomendf =  sqlValue("SELECT COUNT(*) FROM `cache_rating` WHERE `user_id`='" . sql_escape($_REQUEST['userid']) . "'", 0);
 			$content .= '<p><span class="content-title-noshade txt-blue08">'.tr('number_recommendations_given').':</span> <strong>' . $recomendf . '</strong>';
 		
@@ -223,7 +243,7 @@
 							LEFT JOIN	gk_item_waypoint ON gk_item_waypoint.wp = caches.wp_oc
 							LEFT JOIN	gk_item ON gk_item.id = gk_item_waypoint.id AND
 							gk_item.stateid<>1 AND gk_item.stateid<>4 AND gk_item.typeid<>2 AND gk_item.stateid !=5	
-					  WHERE cache_logs.deleted=0 AND `cache_logs`.`user_id`='&1'
+					  WHERE cache_logs.deleted=0 AND cache_logs.hidden=0 AND `cache_logs`.`user_id`='&1'
 					   GROUP BY cache_logs.id
 	                   ORDER BY cache_logs.date_created DESC
 					LIMIT 5", $user_id);
@@ -402,7 +422,7 @@
 	LEFT JOIN	gk_item_waypoint ON gk_item_waypoint.wp = caches.wp_oc
 	LEFT JOIN	gk_item ON gk_item.id = gk_item_waypoint.id AND
 	gk_item.stateid<>1 AND gk_item.stateid<>4 AND gk_item.typeid<>2 AND gk_item.stateid !=5
-					  WHERE cache_logs.deleted=0 AND `caches`.`user_id`='&1'
+					  WHERE cache_logs.deleted=0 AND cache_logs.hidden=0 AND `caches`.`user_id`='&1'
 					  		AND `cache_logs`.`cache_id`=`caches`.`cache_id` 
 							AND `user`.`user_id`=`cache_logs`.`user_id`
 							GROUP BY cache_logs.id
